MapLoaderFailoverTest.testLoadsAll_whenInitialLoaderNodeRemovedWhileLoading

```
org.junit.runners.model.TestTimedOutException: test timed out after 60000 milliseconds
    at java.lang.Object.wait(Native Method)
    at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.pollResponse(InvocationFuture.java:299)
    at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.waitForResponse(InvocationFuture.java:247)
    at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.get(InvocationFuture.java:224)
    at com.hazelcast.util.executor.DelegatingFuture.get(DelegatingFuture.java:72)
    at com.hazelcast.util.executor.DelegatingFuture.get(DelegatingFuture.java:57)
    at com.hazelcast.map.mapstore.MapLoaderFailoverTest.testLoadsAll_whenInitialLoaderNodeRemovedWhileLoading(MapLoaderFailoverTest.java:139)
```

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-OracleJDK1.6/com.hazelcast$hazelcast/531/testReport/junit/com.hazelcast.map.mapstore/MapLoaderFailoverTest/testLoadsAll_whenInitialLoaderNodeRemovedWhileLoading/

```
java.lang.AssertionError: expected:<1> but was:<null>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at com.hazelcast.map.impl.mapstore.MapLoaderFailoverTest.testLoadsAll_whenInitialLoaderNodeRemovedWhileLoading(MapLoaderFailoverTest.java:139)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-OracleJDK8/com.hazelcast$hazelcast/1201/testReport/junit/com.hazelcast.map.impl.mapstore/MapLoaderFailoverTest/testLoadsAll_whenInitialLoaderNodeRemovedWhileLoading/

```
java.lang.AssertionError: expected:<1> but was:<null>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at com.hazelcast.map.impl.mapstore.MapLoaderFailoverTest.testLoadsAll_whenInitialLoaderNodeRemovedWhileLoading(MapLoaderFailoverTest.java:139)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-IbmJDK1.7-freestyle/74/testReport/junit/com.hazelcast.map.impl.mapstore/MapLoaderFailoverTest/testLoadsAll_whenInitialLoaderNodeRemovedWhileLoading/

```
java.lang.AssertionError: expected:<1> but was:<null>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at com.hazelcast.map.impl.mapstore.MapLoaderFailoverTest.testLoadsAll_whenInitialLoaderNodeRemovedWhileLoading(MapLoaderFailoverTest.java:139)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast%20Maintenance/job/Hazelcast-3.maintenance-OracleJDK1.7-freestyle/44/testReport/junit/com.hazelcast.map.impl.mapstore/MapLoaderFailoverTest/testLoadsAll_whenInitialLoaderNodeRemovedWhileLoading/

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-pr-builder/17453/testReport/junit/com.hazelcast.map.impl.mapstore/MapLoaderFailoverTest/testLoadsAll_whenInitialLoaderNodeRemovedWhileLoading/

```
java.lang.AssertionError: expected:<1> but was:<null>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at com.hazelcast.map.impl.mapstore.MapLoaderFailoverTest.testLoadsAll_whenInitialLoaderNodeRemovedWhileLoading(MapLoaderFailoverTest.java:139)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:98)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

--- a/hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/MapLoaderFailoverTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/MapLoaderFailoverTest.java
@@ -136,7 +136,23 @@ public class MapLoaderFailoverTest extends HazelcastTestSupport {
 
         pausingLoader.resume();
 
-        assertEquals(1, asyncVal.get());
+        // workaround for a known MapLoader issue documented in #12384
+        //
+        // in short, there is an edge case in which the get operation is
+        // processed before loading the partition holding the given key
+        // restarts on the previously replica node, after the owner node
+        // died during the load process
+        // for the details, see the issue
+        //
+        // we do this workaround since the goal of the test is to verify
+        // that loadAll() eventually loads all records even if a node
+        // dies in the middle of loading
+        Object getResult = asyncVal.get();
+        if (getResult == null) {
+            getResult = map.get(1);
+        }
+
+        assertEquals(1, getResult);
         assertSizeEventually(MAP_STORE_ENTRY_COUNT, map);
         assertTrue(mapLoader.getLoadedValueCount() >= MAP_STORE_ENTRY_COUNT);
         assertEquals(2, mapLoader.getLoadAllKeysInvocations());

